Mathlib/Order/Minimal.lean

```diff
@@ -1,507 +1,645 @@
+  ext
+  simp_rw [Minimal, hPQ]
+
+theorem maximal_congr (hPQ : ∀ x, P x ↔ Q x) : Maximal P = Maximal Q := by
+  ext
+  simp_rw [Maximal, hPQ]
+
+@[simp] theorem minimal_false : ¬ Minimal (fun _ ↦ False) x := by
+  simp [Minimal]
+
+@[simp] theorem maximal_false : ¬ Maximal (fun _ ↦ False) x := by
+  simp [Maximal]
+
+@[simp] theorem minimal_true : Minimal (fun _ ↦ True) x ↔ IsMin x := by
+  simp [IsMin, Minimal]
+
+@[simp] theorem maximal_true : Maximal (fun _ ↦ True) x ↔ IsMax x :=
+  minimal_true (α := αᵒᵈ)
+
+theorem maximal_true_subtype {x : Subtype P} : Maximal (fun _ ↦ True) x ↔ Maximal P x := by
+  obtain ⟨x, hx⟩ := x
+  simp [Maximal, hx]
+
+theorem minimal_true_subtype {x : Subtype P} : Minimal (fun _ ↦ True) x ↔ Minimal P x := by
+  obtain ⟨x, hx⟩ := x
+  simp [Minimal, hx]
+
+@[simp] theorem minimal_minimal : Minimal (Minimal P) x ↔ Minimal P x :=
+  ⟨fun h ↦ h.prop, fun h ↦ ⟨h, fun _ hy hyx ↦ h.le_of_le hy.prop hyx⟩⟩
+
+@[simp] theorem maximal_maximal : Maximal (Maximal P) x ↔ Maximal P x :=
+  minimal_minimal (α := αᵒᵈ)
+
+/-- If `P` is down-closed, then minimal elements satisfying `P` are exactly the globally minimal
+elements satisfying `P`. -/
+theorem minimal_iff_isMin (hP : ∀ ⦃x y⦄, P y → x ≤ y → P x) : Minimal P x ↔ P x ∧ IsMin x :=
+  ⟨fun h ↦ ⟨h.prop, fun _ h' ↦ h.le_of_le (hP h.prop h') h'⟩, fun h ↦ ⟨h.1, fun _ _  h' ↦ h.2 h'⟩⟩
+
+/-- If `P` is up-closed, then maximal elements satisfying `P` are exactly the globally maximal
+elements satisfying `P`. -/
+theorem maximal_iff_isMax (hP : ∀ ⦃x y⦄, P y → y ≤ x → P x) : Maximal P x ↔ P x ∧ IsMax x :=
+  ⟨fun h ↦ ⟨h.prop, fun _ h' ↦ h.le_of_ge (hP h.prop h') h'⟩, fun h ↦ ⟨h.1, fun _ _  h' ↦ h.2 h'⟩⟩
+
+theorem Minimal.and_right (h : Minimal P x) (hQ : Q x) : Minimal (fun x ↦ (P x ∧ Q x)) x :=
+  ⟨⟨h.prop, hQ⟩, fun _ hy ↦ h.le_of_le hy.1⟩
+
+theorem Minimal.and_left (h : Minimal P x) (hQ : Q x) : Minimal (fun x ↦ (Q x ∧ P x)) x :=
+  ⟨⟨hQ, h.prop⟩, fun _ hy ↦ h.le_of_le hy.2⟩
+
+theorem Maximal.and_right (h : Maximal P x) (hQ : Q x) : Maximal (fun x ↦ (P x ∧ Q x)) x :=
+  ⟨⟨h.prop, hQ⟩, fun _ hy ↦ h.le_of_ge hy.1⟩
+
+theorem Maximal.and_left (h : Maximal P x) (hQ : Q x) : Maximal (fun x ↦ (Q x ∧ P x)) x :=
+  ⟨⟨hQ, h.prop⟩, fun _ hy ↦ h.le_of_ge hy.2⟩
+
+@[simp] theorem minimal_eq_iff : Minimal (· = y) x ↔ x = y := by
+  simp (config := {contextual := true}) [Minimal]
+
+@[simp] theorem maximal_eq_iff : Maximal (· = y) x ↔ x = y := by
+  simp (config := {contextual := true}) [Maximal]
+
+theorem Minimal.or (h : Minimal (fun x ↦ P x ∨ Q x) x) : Minimal P x ∨ Minimal Q x := by
+  obtain ⟨(h | h), hmin⟩ := h
+  · exact .inl ⟨h, fun y hy hyx ↦ hmin (Or.inl hy) hyx⟩
+  exact .inr ⟨h, fun y hy hyx ↦ hmin (Or.inr hy) hyx⟩
+
+theorem Maximal.or (h : Maximal (fun x ↦ P x ∨ Q x) x) : Maximal P x ∨ Maximal Q x :=
+  Minimal.or (α := αᵒᵈ) h
+
+theorem minimal_and_iff_right_of_imp (hPQ : ∀ ⦃x⦄, P x → Q x) :
+    Minimal (fun x ↦ P x ∧ Q x) x ↔ (Minimal P x) ∧ Q x := by
```

Comment:
This lemma is really weird. Without even knowing what the definition of `Minimal` is, the LHS simplifies to
```suggestion
    Minimal P x ↔ (Minimal P x) ∧ Q x := by
```
Maybe just drop it?

Reply:
In the abstract, it looks weird, but the intention here is that the lemma provides a motive for rewriting that would otherwise be annoying to write explicitly, say if you have an goal of the form `_ ∧ Q x ↔ Minimal (fun x ↦ P x ∧ Q x) x` and you want to reduce it to `Q x → ((Minimal P x) ↔ _)`, where `Q` and `P` have long statements.

Reply:
(note that `rw [and_iff_left_of_imp]` wouldn't work in that example because of the binder, and `simp_rw [and_iff_left_of_imp _]` would require an annoying proof term if `P` and `Q` are complicated. )

Reply:
Is `R x ∧ Q x ↔ Minimal (fun x ↦ P x ∧ Q x)` really such a common goal?

Reply:
I'm not making a point purely on the basis of that kind of goal, but that kind of thing has certainly come up for me.

Reply:
I would like to see some concrete examples. I suspect there's a better(/more general?) lemma to be had

Reply:
I think I've answered your question (it provides a motive for a rewrite), and I don't think it makes sense to ask for concrete examples when this is by definition new API. 

If you have an abstract lemma in mind, suggest it!

Reply:
You are free to provide non-minimal concrete examples deep down in one of your repos if that makes it easier for you!

Reply:
Any decision to remove or keep these lemmas is marginally impactful at best, and given that I earnestly anticipate them to be somewhat useful at least in my case, I'd rather just move forward with the PR than continue this discussion.